light-client/src/components/verifier.rs

```diff
@@ -55,75 +56,184 @@ pub trait Verifier: Send + Sync {
-            ProdVotingPowerCalculator::default(),
-            ProdCommitValidator::default(),
-            ProdHasher::default(),
-        )
+impl<P, C, V, H> PredicateVerifier<P, C, V, H>
+where
+    P: VerificationPredicates,
+    C: VotingPowerCalculator,
+    V: CommitValidator,
+    H: Hasher,
+{
+    /// Constructor.
+    pub fn new(predicates: P, voting_power_calculator: C, commit_validator: V, hasher: H) -> Self {
+        Self {
+            predicates,
+            voting_power_calculator,
+            commit_validator,
+            hasher,
+        }
     }
 }
 
-impl Verifier for ProdVerifier {
+impl<P, C, V, H> Verifier for PredicateVerifier<P, C, V, H>
+where
+    P: VerificationPredicates,
+    C: VotingPowerCalculator,
+    V: CommitValidator,
+    H: Hasher,
+{
+    /// Validate the given light block state.
+    ///
+    /// - Ensure the latest trusted header hasn't expired
+    /// - Ensure the header validator hashes match the given validators
+    /// - Ensure the header next validator hashes match the given next
+    ///   validators
+    /// - Additional implementation specific validation via `commit_validator`
+    /// - Check that the untrusted block is more recent than the trusted state
+    /// - If the untrusted block is the very next block after the trusted block,
+    ///   check that their (next) validator sets hashes match.
+    /// - Otherwise, ensure that the untrusted block has a greater height than
+    ///   the trusted block.
+    ///
+    /// **NOTE**: If the untrusted state's `next_validators` field is `None`,
+    /// this will not (and will not be able to) check whether the untrusted
+    /// state's `next_validators_hash` field is valid.
     fn verify(
         &self,
-        untrusted: &LightBlock,
-        trusted: &LightBlock,
+        untrusted: UntrustedBlockState<'_>,
+        trusted: TrustedBlockState<'_>,
         options: &Options,
         now: Time,
     ) -> Verdict {
-        preds::verify(
-            &*self.predicates,
-            &*self.voting_power_calculator,
-            &*self.commit_validator,
-            &*self.hasher,
-            trusted,
-            untrusted,
-            options,
+        // Ensure the latest trusted header hasn't expired
+        verdict!(self.predicates.is_within_trust_period(
+            trusted.header_time,
+            options.trusting_period,
             now,
-        )
-        .into()
+        ));
+
+        // Ensure the header isn't from a future time
+        verdict!(self.predicates.is_header_from_past(
+            untrusted.signed_header.header.time,
+            options.clock_drift,
+            now,
+        ));
+
+        // Ensure the header validator hashes match the given validators
+        verdict!(self.predicates.validator_sets_match(
+            untrusted.validators,
+            untrusted.signed_header.header.validators_hash,
+            &self.hasher,
+        ));
+
+        // TODO(thane): Is this check necessary for IBC?
```

Comment:
@josef-widder We may need your help here.

It seems the IBC handler (ICS 07) has no access to the set of `next_validators` for the untrusted header. For this reason, this field is an `Option` here. When the IBC handler calls into `verify` it does so with this field to `None` (cf. https://github.com/informalsystems/ibc-rs/pull/1321/files#diff-9a833365bf13b4ff1ce2cc757bf18ebabea582b2b97774040e0b7be03fcacbf4R93-R95). This check will effectively be disabled in the context of the IBC handler. This is probably a problem and we need to sort it out.

Reply:
FYI this concern need not block the PR. (Josef will only return from holidays after Sept 27). Maybe @ancazamfir could help us sort out the problem.